### Fase 2 Sul Americana - Quartas de Final

In [1]:
import cartolafc
import pandas as pd
from difflib import get_close_matches
import json
import os

pd.set_option('display.max_columns', 50)            # permite a visualiza??o de 50 colunas do dataframe
pd.options.display.float_format = '{:.2f}'.format   # pandas: para todos os n?meros aparecerem com duas casas decimais

# Cria uma instância da API
api = cartolafc.Api(attempts=5)

2026-05-18 08:19:58,088 - numexpr.utils - INFO - NumExpr defaulting to 8 threads.


### Carregar o arquivo Excel com os classificados e Gerar um Dicion?rio

In [2]:
CAMINHO_ARQUIVO = "1_classificados_oitavas_sula.xlsx"

# Inicializa??o segura
nomes_por_id = {}
dados_torneio_quartas = []

if not os.path.exists(CAMINHO_ARQUIVO):
    print("📁 Arquivo ainda n?o existe — aguardando fim da fase anterior.")
else:
    try:
        # Carrega o arquivo
        df_classificados = pd.read_excel(CAMINHO_ARQUIVO)

        if df_classificados.empty:
            print("📄 Arquivo existe, mas ainda est? vazio. Rodadas da fase anterior n?o conclu?das.")
        else:
            # Filtrar apenas classificados v?lidos
            df_validos = df_classificados[df_classificados["classificado_id"].notnull()].reset_index(drop=True)

            if len(df_validos) < 8:
                print(f"⚠️ Apenas {len(df_validos)} classificados encontrados. Aguardando rodada restante.")
            else:
                # Gerar dicion?rio nome → ID
                nomes_por_id = dict(zip(df_validos["classificado_id"], df_validos["classificado_nome"]))
                display(nomes_por_id)
                print("\n" + "-" * 100 + "\n")

                # Confrontos das quartas com ordem 
                dados_torneio_quartas = [
                    ("Jogo 1 (JG1)", df_validos.loc[0, "classificado_id"]),  # V1
                    ("Jogo 1 (JG1)", df_validos.loc[1, "classificado_id"]),  # V2
                    ("Jogo 2 (JG2)", df_validos.loc[2, "classificado_id"]),  # V3
                    ("Jogo 2 (JG2)", df_validos.loc[3, "classificado_id"]),  # V4
                    ("Jogo 3 (JG3)", df_validos.loc[4, "classificado_id"]),  # V5
                    ("Jogo 3 (JG3)", df_validos.loc[5, "classificado_id"]),  # V6
                    ("Jogo 4 (JG4)", df_validos.loc[6, "classificado_id"]),  # V7
                    ("Jogo 4 (JG4)", df_validos.loc[7, "classificado_id"]),  # V8                    
                ]

                print("✅ Confrontos das Quartas de Final:")
                for linha in dados_torneio_quartas:
                    print(linha)

    except Exception as e:
        print(f"❌ Erro ao processar o arquivo: {e}")


{13951133: 'JUV. KP',
 30267301: 'Máquina Laranjja',
 7017989: 'dasdoresfc',
 1747619: 'JV5 Tricolor Gaúcho',
 18344271: 'FÚRIA LEON',
 18346776: 'AZURRA82',
 24468241: 'Grêmio imortal 37',
 19033717: 'Mau Humor F.C.'}


----------------------------------------------------------------------------------------------------

✅ Confrontos das Quartas de Final:
('Jogo 1 (JG1)', 13951133)
('Jogo 1 (JG1)', 30267301)
('Jogo 2 (JG2)', 7017989)
('Jogo 2 (JG2)', 1747619)
('Jogo 3 (JG3)', 18344271)
('Jogo 3 (JG3)', 18346776)
('Jogo 4 (JG4)', 24468241)
('Jogo 4 (JG4)', 19033717)


In [3]:
# Criar DataFrame base
dados_torneio_quartas = pd.DataFrame(dados_torneio_quartas, columns=["Jogo", "ID do Time"])

# Adicionar Nome do Time usando o dicion?rio
dados_torneio_quartas["Nome do Time"] = dados_torneio_quartas["ID do Time"].map(nomes_por_id)

# Adicionar ID no Grupo
dados_torneio_quartas["ID no Grupo"] = dados_torneio_quartas.groupby("Jogo").cumcount() + 1
dados_torneio_quartas["ID no Grupo"] = dados_torneio_quartas["ID no Grupo"].astype(str) + "_" + dados_torneio_quartas["Jogo"].str[-2]

# Reorganizar colunas
df_torneio_quartas_sula = dados_torneio_quartas[["Jogo", "ID do Time", "Nome do Time", "ID no Grupo"]]

df_sula_jogo_1 = df_torneio_quartas_sula[df_torneio_quartas_sula["Jogo"] == "Jogo 1 (JG1)"]
df_sula_jogo_2 = df_torneio_quartas_sula[df_torneio_quartas_sula["Jogo"] == "Jogo 2 (JG2)"]
df_sula_jogo_3 = df_torneio_quartas_sula[df_torneio_quartas_sula["Jogo"] == "Jogo 3 (JG3)"]
df_sula_jogo_4 = df_torneio_quartas_sula[df_torneio_quartas_sula["Jogo"] == "Jogo 4 (JG4)"]


# Lista de grupos
grupos = {
    "Jogo 1 (JG1)": df_sula_jogo_1,
    "Jogo 2 (JG2)": df_sula_jogo_2,
    "Jogo 3 (JG3)": df_sula_jogo_3,
    "Jogo 4 (JG4)": df_sula_jogo_4
}

display(df_sula_jogo_1, df_sula_jogo_2, df_sula_jogo_3, df_sula_jogo_4)

,Jogo,ID do Time,Nome do Time,ID no Grupo
0,Jogo 1 (JG1),13951133,JUV. KP,1_1
1,Jogo 1 (JG1),30267301,Máquina Laranjja,2_1


,Jogo,ID do Time,Nome do Time,ID no Grupo
2,Jogo 2 (JG2),7017989,dasdoresfc,1_2
3,Jogo 2 (JG2),1747619,JV5 Tricolor Gaúcho,2_2


,Jogo,ID do Time,Nome do Time,ID no Grupo
4,Jogo 3 (JG3),18344271,FÚRIA LEON,1_3
5,Jogo 3 (JG3),18346776,AZURRA82,2_3


,Jogo,ID do Time,Nome do Time,ID no Grupo
6,Jogo 4 (JG4),24468241,Grêmio imortal 37,1_4
7,Jogo 4 (JG4),19033717,Mau Humor F.C.,2_4


### Defini??o dos Confrontos das 2 Rodadas da Fase 2 da Sul Americana (Rodada 15 A 35)

In [4]:
# Rodada 1 - Fase 2 Sul Americana (Equivalente a 34? Rodada do Campeonato Brasileiro) 
confrontos_1a_rodada = [
    # Jogo 1 (JG1)
    ("Jogo 1 (JG1)", "1_1", "2_1"),    

    # Jogo 2 (JG2)
    ("Jogo 2 (JG2)", "1_2", "2_2"),   

    # Jogo 3 (JG3)
    ("Jogo 3 (JG3)", "1_3", "2_3"),

    # Jogo 4 (JG4)
    ("Jogo 4 (JG4)", "1_4", "2_4")
]

# Rodada 2 - Fase 2 Sul Americana (Equivalente a 35? Rodada do Campeonato Brasileiro)
confrontos_2a_rodada = [
    # Jogo 1 (JG1)
    ("Jogo 1 (JG1)", "2_1", "1_1"),
    
    # Jogo 2 (JG2)
    ("Jogo 2 (JG2)", "2_2", "1_2"),

    # Jogo 3 (JG3)
    ("Jogo 3 (JG3)", "2_3", "1_3"),

    # Jogo 4 (JG4)
    ("Jogo 4 (JG4)", "2_4", "1_4") 
]


In [5]:
# Transformar em DataFrame
df_confrontos = pd.DataFrame(confrontos_1a_rodada, columns=["Grupo", "Mandante_ID", "Visitante_ID"])

# Junta com df_torneio para buscar dados dos mandantes
df_mandantes = df_torneio_quartas_sula.rename(columns={
    "ID no Grupo": "Mandante_ID",
    "Nome do Time": "Mandante_Nome",
    "ID do Time": "Mandante_ID_Time"
})[["Jogo", "Mandante_ID", "Mandante_Nome", "Mandante_ID_Time"]]

# Junta com df_torneio para buscar dados dos visitantes
df_visitantes = df_torneio_quartas_sula.rename(columns={
    "ID no Grupo": "Visitante_ID",
    "Nome do Time": "Visitante_Nome",
    "ID do Time": "Visitante_ID_Time"    
})[["Jogo", "Visitante_ID", "Visitante_Nome", "Visitante_ID_Time"]]

display(df_confrontos)

,Grupo,Mandante_ID,Visitante_ID
0,Jogo 1 (JG1),1_1,2_1
1,Jogo 2 (JG2),1_2,2_2
2,Jogo 3 (JG3),1_3,2_3
3,Jogo 4 (JG4),1_4,2_4


In [6]:
# Transformar em DataFrame
df_confrontos = pd.DataFrame(confrontos_1a_rodada, columns=["Jogo", "Mandante_ID", "Visitante_ID"])
df_confrontos["Rodada"] = 15
df_rodada_15 = df_confrontos.merge(df_mandantes, on=["Jogo", "Mandante_ID"])
df_rodada_15 = df_rodada_15.merge(df_visitantes, on=["Jogo", "Visitante_ID"])

# Transformar em DataFrame
df_confrontos_2 = pd.DataFrame(confrontos_2a_rodada, columns=["Jogo", "Mandante_ID", "Visitante_ID"])
df_confrontos_2["Rodada"] = 16
df_rodada_16 = df_confrontos_2.merge(df_mandantes, on=["Jogo", "Mandante_ID"])
df_rodada_16 = df_rodada_16.merge(df_visitantes, on=["Jogo", "Visitante_ID"])

display(df_rodada_15)

,Jogo,Mandante_ID,Visitante_ID,Rodada,Mandante_Nome,Mandante_ID_Time,Visitante_Nome,Visitante_ID_Time
0,Jogo 1 (JG1),1_1,2_1,15,JUV. KP,13951133,Máquina Laranjja,30267301
1,Jogo 2 (JG2),1_2,2_2,15,dasdoresfc,7017989,JV5 Tricolor Gaúcho,1747619
2,Jogo 3 (JG3),1_3,2_3,15,FÚRIA LEON,18344271,AZURRA82,18346776
3,Jogo 4 (JG4),1_4,2_4,15,Grêmio imortal 37,24468241,Mau Humor F.C.,19033717


In [7]:
df_rodadas = pd.concat([
    df_rodada_15,
    df_rodada_16
], ignore_index=True)

# Ajustar a numeração da rodada para refletir as rodadas do Cartola (15 e 16)
df_rodadas["Rodada"] = df_rodadas["Rodada"]

df_rodadas.to_excel("2_confrontos_quartas_sula.xlsx", index=False)

# Exibir os confrontos da fase 4 (Quartas de Final)
display(df_rodadas.head(8)) 

,Jogo,Mandante_ID,Visitante_ID,Rodada,Mandante_Nome,Mandante_ID_Time,Visitante_Nome,Visitante_ID_Time
0,Jogo 1 (JG1),1_1,2_1,15,JUV. KP,13951133,Máquina Laranjja,30267301
1,Jogo 2 (JG2),1_2,2_2,15,dasdoresfc,7017989,JV5 Tricolor Gaúcho,1747619
2,Jogo 3 (JG3),1_3,2_3,15,FÚRIA LEON,18344271,AZURRA82,18346776
3,Jogo 4 (JG4),1_4,2_4,15,Grêmio imortal 37,24468241,Mau Humor F.C.,19033717
4,Jogo 1 (JG1),2_1,1_1,16,Máquina Laranjja,30267301,JUV. KP,13951133
5,Jogo 2 (JG2),2_2,1_2,16,JV5 Tricolor Gaúcho,1747619,dasdoresfc,7017989
6,Jogo 3 (JG3),2_3,1_3,16,AZURRA82,18346776,FÚRIA LEON,18344271
7,Jogo 4 (JG4),2_4,1_4,16,Mau Humor F.C.,19033717,Grêmio imortal 37,24468241


In [8]:
# Criar lista de dicion?rios no formato desejado
confrontos_js_quartas = []

for _, row in df_rodadas.iterrows():
    confronto = {
        "jogo": row["Jogo"],
        "rodada": int(row["Rodada"]),
        "mandante": {
            "id": int(row["Mandante_ID_Time"]),
            "nome": row["Mandante_Nome"]
        },
        "visitante": {
            "id": int(row["Visitante_ID_Time"]),
            "nome": row["Visitante_Nome"]        }
    }
    confrontos_js_quartas.append(confronto)

# Converter para JSON formatado
json_str = json.dumps(confrontos_js_quartas, indent=2, ensure_ascii=False)

# Salvar como arquivo JS com uma vari?vel global
with open("2_confrontos_quartas_sula.js", "w", encoding="utf-8") as f:
    f.write("const confrontos_quartas_sula = ")
    f.write(json_str)
    f.write(";")

In [9]:
def exibir_confrontos(df_rodadas, rodada=None, jogo=None):

    colunas = ["Rodada", "Jogo", "Mandante_Nome", "Visitante_Nome"]
    df_filtrado = df_rodadas.copy()

    df_filtrado["Rodada"] = df_filtrado["Rodada"].astype(str) + "ª Rodada"    

    if rodada is not None:
        df_filtrado = df_filtrado[df_filtrado["Rodada"] == rodada]

    if jogo is not None:
        df_filtrado = df_filtrado[df_filtrado["Jogo"] == jogo]

    return df_filtrado[colunas].sort_values(by=["Jogo", "Rodada"])

In [10]:
jogo = 1

# Exibir todos os confrontos do Jogo 2
display(exibir_confrontos(df_rodadas, jogo= f"Jogo {jogo} (JG{jogo})").head(8))

,Rodada,Jogo,Mandante_Nome,Visitante_Nome
0,15ª Rodada,Jogo 1 (JG1),JUV. KP,Máquina Laranjja
4,16ª Rodada,Jogo 1 (JG1),Máquina Laranjja,JUV. KP


In [11]:
# def campeonato_comecou(api, ids_times):
#     """Verifica se o campeonato j? come?ou observando a pontua??o na 1? rodada."""
#     for time_id in ids_times.values():
#         try:
#             pontuacao = api.time(time_id=time_id, rodada=15).ultima_pontuacao
#             if pontuacao is not None:
#                 return True
#         except cartolafc.errors.CartolaFCError:
#             continue
#     return False

# def obter_pontuacao_por_rodada(api, time_id, rodada_atual):
#     """Obt?m a pontua??o do time em cada rodada at? a rodada atual."""
#     pontuacoes = {}
#     for rodada in range(15, rodada_atual):
#         try:
#             time_rodada = api.time(time_id=time_id, rodada=rodada)
#             pontuacoes[rodada] = time_rodada.ultima_pontuacao
#         except cartolafc.errors.CartolaFCError as e:
#             print(f"Erro ao acessar pontua??o da rodada {rodada} para o time {time_id}: {e}")
#             pontuacoes[rodada] = None
#     return pontuacoes



# def gerar_df_pontuacoes(api, ids_times, rodada_inicio=15):
#     rodada_atual = api.mercado().rodada_atual

#     if not campeonato_comecou(api, ids_times):
#         print("📌 A fase ainda n?o come?ou. Criando estrutura com placeholders.")
#         df = pd.DataFrame(index=ids_times.keys(), columns=[f'Rodada {i}' for i in range(rodada_inicio, rodada_inicio + 2)])
#         df[:] = None  # ou 0 se preferir
#         return df

#     # Campeonato come?ou
#     df = pd.DataFrame()
#     for nome, time_id in ids_times.items():
#         pontuacoes = obter_pontuacao_por_rodada(api, time_id, rodada_atual)
#         df[nome] = pd.Series(pontuacoes)

#     df = df.transpose()

#     # Ajustar nome das colunas para "Rodada X"
#     colunas_numeradas = df.columns.tolist()
#     df.columns = [f'Rodada {i}' for i in colunas_numeradas]

#     # Adicionar a linha com o time l?der por rodada (opcional)
#     if not df.empty:
#         df.loc['Lider_Rodada'] = df.idxmax()

#     return df


In [12]:
import requests
import time
import sys
from pathlib import Path

FASE_INICIO = 15
FASE_LIMITE = 16
PER_REQ_SLEEP = 1.0

ROOT = Path.cwd().parents[1]
if str(ROOT) not in sys.path:
    sys.path.insert(0, str(ROOT))

import scripts.parciais as parciais
parciais.HEADERS = globals().get("HEADERS", {})
fetch_pontuados = parciais.fetch_pontuados
fetch_time_payload = parciais.fetch_time_payload
clubes_que_ja_jogaram = parciais.clubes_que_ja_jogaram
calcular_parcial_time = parciais.calcular_parcial_time

sess = requests.Session()
sess.headers.update({
    "User-Agent": "Mozilla/5.0",
    "Accept": "application/json, text/plain, */*",
})


def http_status_e_rodada():
    r = sess.get("https://api.cartola.globo.com/mercado/status", timeout=20)
    r.raise_for_status()
    data = r.json()
    return int(data.get("status_mercado", 0)), int(data.get("rodada_atual", 0))


def _coerce_float(value):
    try:
        if value is None:
            return None
        return float(value)
    except Exception:
        return None


def extrair_pontuacao_payload(payload):
    if not isinstance(payload, dict):
        return None

    candidatos = []
    for chave in ("pontos", "pontuacao"):
        candidatos.append(payload.get(chave))

    time_info = payload.get("time", {})
    if isinstance(time_info, dict):
        for chave in ("pontos", "pontuacao"):
            candidatos.append(time_info.get(chave))

    pontos_info = payload.get("pontos")
    if isinstance(pontos_info, dict):
        for chave in ("rodada", "total", "valor"):
            candidatos.append(pontos_info.get(chave))

    for candidato in candidatos:
        valor = _coerce_float(candidato)
        if valor is not None:
            return valor

    return None


def obter_pontuacao_rodada_fechada(api, time_id, rodada):
    payload = fetch_time_payload(int(time_id), int(rodada))
    pontuacao_payload = extrair_pontuacao_payload(payload)
    if pontuacao_payload is not None:
        return pontuacao_payload

    time_rodada = api.time(time_id=int(time_id), rodada=int(rodada))
    for atributo in ("pontos", "pontuacao", "ultima_pontuacao"):
        valor = _coerce_float(getattr(time_rodada, atributo, None))
        if valor is not None:
            return valor

    return None


def gerar_df_pontuacoes(api, ids_times):
    global status_http, rodada_http, rodada_api, rod_ref, FASE_FIM, RODADAS_CONCLUIDAS_FIM

    colunas_fase = [f"Rodada {i}" for i in range(FASE_INICIO, FASE_LIMITE + 1)]

    try:
        status_http, rodada_http = http_status_e_rodada()
    except Exception:
        status_http, rodada_http = 0, 0

    try:
        rodada_api = int(api.mercado().rodada_atual)
    except Exception:
        rodada_api = int(rodada_http or 0)

    rod_ref = int(rodada_http or rodada_api or 0)

    if status_http == 2:
        FASE_FIM = max(FASE_INICIO, min(FASE_LIMITE, rod_ref))
        RODADAS_CONCLUIDAS_FIM = FASE_FIM - 1
    elif status_http == 1:
        FASE_FIM = max(FASE_INICIO, min(FASE_LIMITE, max(rod_ref - 1, FASE_INICIO)))
        RODADAS_CONCLUIDAS_FIM = FASE_FIM
    else:
        FASE_FIM = max(FASE_INICIO, min(FASE_LIMITE, rod_ref if rod_ref else FASE_INICIO))
        RODADAS_CONCLUIDAS_FIM = FASE_FIM

    print(
        f"Status={status_http} | rodada_http={rodada_http} | rodada_api={rodada_api} | "
        f"FASE_INICIO={FASE_INICIO} | FASE_FIM={FASE_FIM} | fechadas ate {RODADAS_CONCLUIDAS_FIM}"
    )

    dados = {}
    for nome, time_id in ids_times.items():
        pontuacoes = {}
        if RODADAS_CONCLUIDAS_FIM >= FASE_INICIO:
            for rodada in range(FASE_INICIO, RODADAS_CONCLUIDAS_FIM + 1):
                try:
                    pontuacoes[rodada] = obter_pontuacao_rodada_fechada(api, time_id, rodada)
                except Exception as e:
                    print(f"Erro ao acessar pontuacao da rodada {rodada} para o time {time_id}: {e}")
                    pontuacoes[rodada] = None
        dados[nome] = {f"Rodada {rodada}": pontuacoes.get(rodada) for rodada in range(FASE_INICIO, FASE_LIMITE + 1)}

    df = pd.DataFrame.from_dict(dados, orient="index")
    df = df.reindex(columns=colunas_fase)
    df = df.apply(pd.to_numeric, errors="coerce")

    col_atual = f"Rodada {rod_ref}"
    if status_http == 2 and FASE_INICIO <= rod_ref <= FASE_LIMITE:
        print(f"Rodada {rod_ref} em andamento: aplicando parciais")
        mapa_pontuados = fetch_pontuados()
        if mapa_pontuados:
            clubes_jogaram = clubes_que_ja_jogaram(rod_ref)
            for nome_time, time_id in ids_times.items():
                try:
                    total = calcular_parcial_time(int(time_id), rod_ref, mapa_pontuados, clubes_jogaram)
                    if col_atual not in df.columns:
                        df[col_atual] = pd.NA
                    df.loc[nome_time, col_atual] = round(total, 2)
                except Exception as e:
                    print(f"Erro ao calcular parcial da rodada {rod_ref} para o time {time_id}: {e}")
                time.sleep(PER_REQ_SLEEP)
        else:
            print("Parciais indisponiveis no momento.")
    else:
        print("Sem parciais para aplicar nesta fase.")

    colunas_com_dados = [col for col in df.columns if df[col].notna().any()]
    if colunas_com_dados:
        df.loc["Lider_Rodada", colunas_com_dados] = df[colunas_com_dados].idxmax()
    if not colunas_com_dados:
        df.loc["Lider_Rodada"] = None

    return df


In [13]:
ids_times = {v: k for k, v in nomes_por_id.items()}
df_pontuacoes = gerar_df_pontuacoes(api, ids_times)

if df_pontuacoes.empty:
    print("⚠️ Sem pontua??es dispon?veis ainda.")
else:
    display(df_pontuacoes.T)

Status=2 | rodada_http=16 | rodada_api=16 | FASE_INICIO=15 | FASE_FIM=16 | fechadas ate 15


Rodada 16 em andamento: aplicando parciais


C:\Users\ferna\AppData\Local\Temp\ipykernel_10416\2059488124.py:155: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'Mau Humor F.C.' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  df.loc["Lider_Rodada", colunas_com_dados] = df[colunas_com_dados].idxmax()
C:\Users\ferna\AppData\Local\Temp\ipykernel_10416\2059488124.py:155: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'JV5 Tricolor Gaúcho' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  df.loc["Lider_Rodada", colunas_com_dados] = df[colunas_com_dados].idxmax()


,JUV. KP,Máquina Laranjja,dasdoresfc,JV5 Tricolor Gaúcho,FÚRIA LEON,AZURRA82,Grêmio imortal 37,Mau Humor F.C.,Lider_Rodada
Rodada 15,55.05,37.83,55.90,55.90,43.73,43.82,67.55,68.35,Mau Humor F.C.
Rodada 16,75.88,56.93,82.25,103.39,66.58,59.61,88.68,74.28,JV5 Tricolor Gaúcho


In [14]:
# def classificacao_por_grupo(df_rodadas, df_pontuacoes):

#     df_pontuacoes_times = df_pontuacoes.drop(index='Lider_Rodada', errors='ignore')
#     estatisticas = {}

#     for _, confronto in df_rodadas.iterrows():
#         rodada = confronto["Rodada"]
#         mandante = confronto["Mandante_Nome"]
#         visitante = confronto["Visitante_Nome"]
#         jogo = confronto["Jogo"]
#         coluna_rodada = f"Rodada {rodada}"

#         if mandante not in df_pontuacoes_times.index or visitante not in df_pontuacoes_times.index:
#             continue
#         if coluna_rodada not in df_pontuacoes_times.columns:
#             continue

#         pontos_mandante = df_pontuacoes_times.at[mandante, coluna_rodada]
#         pontos_visitante = df_pontuacoes_times.at[visitante, coluna_rodada]

#         # Ignorar confrontos ainda n?o disputados (com pontua??o 0 ou ausente)
#         if (
#             pd.isnull(pontos_mandante) or pd.isnull(pontos_visitante) or
#             (pontos_mandante == 0 and pontos_visitante == 0)
#         ):
#             continue

#         for time in [mandante, visitante]:
#             if jogo not in estatisticas:
#                 estatisticas[jogo] = {}
#             if time not in estatisticas[jogo]:
#                 estatisticas[jogo][time] = {
#                     "Pontos": 0, "Vit?rias": 0, "Empates": 0, "Derrotas": 0,
#                     "Total_Cartola": 0,
#                     "Cartola_Sofrido": 0
#                 }

#         # Atualizar estat?sticas do jogo
#         estatisticas[jogo][mandante]["Total_Cartola"] += pontos_mandante
#         estatisticas[jogo][mandante]["Cartola_Sofrido"] += pontos_visitante

#         estatisticas[jogo][visitante]["Total_Cartola"] += pontos_visitante
#         estatisticas[jogo][visitante]["Cartola_Sofrido"] += pontos_mandante

#         if pontos_mandante > pontos_visitante:
#             estatisticas[jogo][mandante]["Pontos"] += 3
#             estatisticas[jogo][mandante]["Vit?rias"] += 1
#             estatisticas[jogo][visitante]["Derrotas"] += 1
#         elif pontos_mandante < pontos_visitante:
#             estatisticas[jogo][visitante]["Pontos"] += 3
#             estatisticas[jogo][visitante]["Vit?rias"] += 1
#             estatisticas[jogo][mandante]["Derrotas"] += 1
#         else:
#             estatisticas[jogo][mandante]["Pontos"] += 1
#             estatisticas[jogo][visitante]["Pontos"] += 1
#             estatisticas[jogo][mandante]["Empates"] += 1
#             estatisticas[jogo][visitante]["Empates"] += 1

#     # Gerar DataFrame final
#     df_resultado = pd.concat([
#         pd.DataFrame({
#             "Jogo": jogo,
#             "Nome do Time": list(times.keys()),
#             "Pontos": [stats["Pontos"] for stats in times.values()],
#             "Vit?rias": [stats["Vit?rias"] for stats in times.values()],
#             "Empates": [stats["Empates"] for stats in times.values()],
#             "Derrotas": [stats["Derrotas"] for stats in times.values()],
#             "Total Cartola": [stats["Total_Cartola"] for stats in times.values()],
#             "Cartola Sofrido": [stats["Cartola_Sofrido"] for stats in times.values()],
#             "Saldo Cartola": [
#                 stats["Total_Cartola"] - stats["Cartola_Sofrido"] for stats in times.values()
#             ]
#         }) for jogo, times in estatisticas.items()
#     ], ignore_index=True)

#     # Ordenar e adicionar posi??o
#     df_resultado = df_resultado.sort_values(
#         by=["Jogo", "Pontos", "Vit?rias", "Total Cartola", "Saldo Cartola"],
#         ascending=[True, False, False, False, False]
#     )

#     df_resultado["Posi??o"] = df_resultado.groupby("Jogo").cumcount() + 1

#     df_resultado_por_jogo = {
#         jogo: df_resultado[df_resultado["Jogo"] == jogo] for jogo in df_resultado["Jogo"].unique()
#     }

#     return df_resultado, df_resultado_por_jogo


In [15]:
def classificacao_por_grupo(df_rodadas, df_pontuacoes, rodada_atual=None):
    """
    Gera classifica??o por grupo (ou jogo), exibindo times zerados
    se a fase ainda n?o come?ou (sem pontua??es nas rodadas da fase).
    """
    df_pontuacoes_times = df_pontuacoes.drop(index='Lider_Rodada', errors='ignore')
    estatisticas = {}

    # 🔹 Descobre rodadas da fase
    rodadas_fase = sorted(df_rodadas["Rodada"].unique())
    rodadas_validas = [r for r in rodadas_fase if f"Rodada {r}" in df_pontuacoes.columns]

    # 🔹 Se rodada_atual for anterior ? primeira rodada da fase → retorna tabela zerada
    if rodada_atual is not None and rodada_atual < min(rodadas_fase):
        print(f"📅 Fase ainda n?o come?ou (rodada atual {rodada_atual}). Exibindo zerado.")
        return montar_tabela_zerada(df_rodadas)

    # 🔹 Verifica se h? alguma pontua??o v?lida
    tem_pontuacao = any(
        pd.notna(df_pontuacoes[f"Rodada {r}"]).any() for r in rodadas_validas
    )
    if not tem_pontuacao:
        print("📊 Nenhuma pontua??o registrada nesta fase. Exibindo classifica??o inicial zerada.")
        return montar_tabela_zerada(df_rodadas)

    # 🔹 Caso existam pontua??es, gera normalmente
    for _, confronto in df_rodadas.iterrows():
        rodada = confronto.get("Rodada")
        mandante = confronto.get("Mandante_Nome")
        visitante = confronto.get("Visitante_Nome")
        jogo = confronto.get("Jogo")
        coluna_rodada = f"Rodada {rodada}"

        if coluna_rodada not in df_pontuacoes_times.columns:
            continue

        pontos_mandante = df_pontuacoes_times.at[mandante, coluna_rodada]
        pontos_visitante = df_pontuacoes_times.at[visitante, coluna_rodada]
        pontos_mandante = 0 if pd.isnull(pontos_mandante) else pontos_mandante
        pontos_visitante = 0 if pd.isnull(pontos_visitante) else pontos_visitante

        if jogo not in estatisticas:
            estatisticas[jogo] = {}
        for time in [mandante, visitante]:
            estatisticas[jogo].setdefault(time, {
                "Pontos": 0, "Vit?rias": 0, "Empates": 0, "Derrotas": 0,
                "Total_Cartola": 0, "Cartola_Sofrido": 0
            })

        estatisticas[jogo][mandante]["Total_Cartola"] += pontos_mandante
        estatisticas[jogo][mandante]["Cartola_Sofrido"] += pontos_visitante
        estatisticas[jogo][visitante]["Total_Cartola"] += pontos_visitante
        estatisticas[jogo][visitante]["Cartola_Sofrido"] += pontos_mandante

        # Pula se ainda n?o houve pontua??o real
        if pontos_mandante == 0 and pontos_visitante == 0:
            continue

        if pontos_mandante > pontos_visitante:
            estatisticas[jogo][mandante]["Pontos"] += 3
            estatisticas[jogo][mandante]["Vit?rias"] += 1
            estatisticas[jogo][visitante]["Derrotas"] += 1
        elif pontos_mandante < pontos_visitante:
            estatisticas[jogo][visitante]["Pontos"] += 3
            estatisticas[jogo][visitante]["Vit?rias"] += 1
            estatisticas[jogo][mandante]["Derrotas"] += 1
        else:
            estatisticas[jogo][mandante]["Pontos"] += 1
            estatisticas[jogo][visitante]["Pontos"] += 1
            estatisticas[jogo][mandante]["Empates"] += 1
            estatisticas[jogo][visitante]["Empates"] += 1

    # 🔹 Monta dataframe final
    df_resultado = pd.concat([
        pd.DataFrame({
            "Jogo": jogo,
            "Nome do Time": list(times.keys()),
            "Pontos": [stats["Pontos"] for stats in times.values()],
            "Vit?rias": [stats["Vit?rias"] for stats in times.values()],
            "Empates": [stats["Empates"] for stats in times.values()],
            "Derrotas": [stats["Derrotas"] for stats in times.values()],
            "Total Cartola": [stats["Total_Cartola"] for stats in times.values()],
            "Cartola Sofrido": [stats["Cartola_Sofrido"] for stats in times.values()],
            "Saldo Cartola": [
                stats["Total_Cartola"] - stats["Cartola_Sofrido"] for stats in times.values()
            ]
        })
        for jogo, times in estatisticas.items()
    ], ignore_index=True)

    df_resultado = df_resultado.sort_values(
        by=["Jogo", "Pontos", "Vit?rias", "Total Cartola", "Saldo Cartola"],
        ascending=[True, False, False, False, False]
    )
    df_resultado["Posi??o"] = df_resultado.groupby("Jogo").cumcount() + 1

    df_resultado_por_jogo = {
        jogo: df_resultado[df_resultado["Jogo"] == jogo] for jogo in df_resultado["Jogo"].unique()
    }

    return df_resultado, df_resultado_por_jogo


# 🔧 Fun??o auxiliar
def montar_tabela_zerada(df_rodadas):
    """Cria tabela zerada com base nos confrontos informados."""
    estatisticas = {
        jogo: {
            row["Mandante_Nome"]: {"Pontos": 0, "Vit?rias": 0, "Empates": 0, "Derrotas": 0,
                                   "Total_Cartola": 0, "Cartola_Sofrido": 0},
            row["Visitante_Nome"]: {"Pontos": 0, "Vit?rias": 0, "Empates": 0, "Derrotas": 0,
                                    "Total_Cartola": 0, "Cartola_Sofrido": 0}
        }
        for jogo, row in df_rodadas.groupby("Jogo").first().iterrows()
    }

    df_resultado = pd.concat([
        pd.DataFrame({
            "Jogo": jogo,
            "Nome do Time": list(times.keys()),
            "Pontos": [0, 0],
            "Vit?rias": [0, 0],
            "Empates": [0, 0],
            "Derrotas": [0, 0],
            "Total Cartola": [0, 0],
            "Cartola Sofrido": [0, 0],
            "Saldo Cartola": [0, 0],
        })
        for jogo, times in estatisticas.items()
    ], ignore_index=True)

    df_resultado["Posi??o"] = 0
    df_resultado_por_jogo = {
        jogo: df_resultado[df_resultado["Jogo"] == jogo] for jogo in df_resultado["Jogo"].unique()
    }
    return df_resultado, df_resultado_por_jogo


In [16]:
# Padroniza e exibe as pontua??es se existirem
try:
    if 'df_pontuacoes' not in globals() or df_pontuacoes.empty:
        print("📌 Nenhuma pontua??o dispon?vel — provavelmente ainda n?o h? classificados definidos.")
    else:
        # Padroniza os nomes no df_rodadas
        df_rodadas["Mandante_Nome"] = df_rodadas["Mandante_Nome"].str.strip()
        df_rodadas["Visitante_Nome"] = df_rodadas["Visitante_Nome"].str.strip()

        # Padroniza os ?ndices do df_pontuacoes
        df_pontuacoes.index = df_pontuacoes.index.str.strip()

        # Exibe
        display(df_pontuacoes)

except NameError as e:
    print("❌ Vari?vel df_rodadas ou df_pontuacoes n?o est? definida ainda.")
except Exception as e:
    print(f"❌ Erro inesperado: {e}")


,Rodada 15,Rodada 16
JUV. KP,55.05,75.88
Máquina Laranjja,37.83,56.93
dasdoresfc,55.90,82.25
JV5 Tricolor Gaúcho,55.90,103.39
FÚRIA LEON,43.73,66.58
AZURRA82,43.82,59.61
Grêmio imortal 37,67.55,88.68
Mau Humor F.C.,68.35,74.28
Lider_Rodada,Mau Humor F.C.,JV5 Tricolor Gaúcho


In [17]:
try:
    # Verifica se as vari?veis existem e est?o v?lidas
    if (
        'df_rodadas' not in globals() or df_rodadas.empty or
        'df_pontuacoes' not in globals() or df_pontuacoes.empty
    ):
        print("⚠️ Rodadas ou pontua??es ainda n?o dispon?veis — classifica??o n?o ser? gerada.")
    
    elif 'classificacao_por_grupo' not in globals():
        print("⚠️ Fun??o 'classificacao_por_grupo' n?o est? definida.")
    
    else:
        # Gerar a classifica??o da fase 4
        df_resultado, df_resultado_por_jogo = classificacao_por_grupo(
            df_rodadas, df_pontuacoes
        )

        # Salvar cada grupo em uma aba do Excel
        with pd.ExcelWriter("2_classificacao_por_jogo_quartas_sula.xlsx") as writer:
            for jogo, df in df_resultado_por_jogo.items():
                df.to_excel(writer, sheet_name=jogo, index=False)

        # Exibir a classifica??o geral
        df_resultado_jogo_1 = df_resultado[df_resultado["Jogo"] == "Jogo 1 (JG1)"]
        df_resultado_jogo_2 = df_resultado[df_resultado["Jogo"] == "Jogo 2 (JG2)"]

        display(df_resultado_jogo_1, df_resultado_jogo_2)

except Exception as e:
    print(f"❌ Erro ao gerar a classifica??o: {e}")


,Jogo,Nome do Time,Pontos,Vit?rias,Empates,Derrotas,Total Cartola,Cartola Sofrido,Saldo Cartola,Posi??o
0,Jogo 1 (JG1),JUV. KP,6,2,0,0,130.93,94.76,36.17,1
1,Jogo 1 (JG1),Máquina Laranjja,0,0,0,2,94.76,130.93,-36.17,2


,Jogo,Nome do Time,Pontos,Vit?rias,Empates,Derrotas,Total Cartola,Cartola Sofrido,Saldo Cartola,Posi??o
3,Jogo 2 (JG2),JV5 Tricolor Gaúcho,4,1,1,0,159.29,138.15,21.14,1
2,Jogo 2 (JG2),dasdoresfc,1,0,1,1,138.15,159.29,-21.14,2


{18421230: 'Gig@ntte',
 13951133: 'KING LEONN',
 18344271: 'Laranjja Mecannica',
 24468241: 'Gr?mio imortal 37',
 25751748: 'MauHumor F.C.',
 479510: 'TEAM LOPES 99',
 47543456: 'Tabajara de Inhaua FC2',
 44810918: 'lsauer fc'

In [18]:
# Verifica se a vari?vel df_resultado_por_grupo existe e est? populada
if 'df_resultado_por_jogo' in locals() and df_resultado_por_jogo:
    # Criar estrutura em formato de dicion?rio para JSON/JS
    classificacao_quartas_sula = {}

    for jogo, df in df_resultado_por_jogo.items():
        classificacao_quartas_sula[jogo] = []

        # Remove duplicatas
        df = df.drop_duplicates(subset=["Nome do Time"])

        for _, row in df.iterrows():
            classificacao_quartas_sula[jogo].append({
                "posicao": int(row["Posi??o"]),
                "nome": row["Nome do Time"],
                "pontos": int(row["Pontos"]),
                "vitorias": int(row["Vit?rias"]),
                "empates": int(row["Empates"]),
                "derrotas": int(row["Derrotas"]),
                "totalCartola": float(row["Total Cartola"]),
                "cartolaSofrido": float(row["Cartola Sofrido"]),
                "saldoCartola": float(row["Saldo Cartola"])
            })


    # Converter para JSON formatado
    json_str = json.dumps(classificacao_quartas_sula, indent=2, ensure_ascii=False)

    # Salvar como arquivo JS com uma vari?vel global
    with open("2_classificacao_por_jogo_quartas_sula.js", "w", encoding="utf-8") as f:
        f.write("const classificacao_quartas_sula = ")
        f.write(json_str)
        f.write(";")

    print("✅ Arquivo JS salvo com sucesso.")

else:
    print("⚠️ Classifica??o das Quartas da Sula indispon?veis. Criando placeholders.")

    # ⚠️ Classifica??o da fase 4 indispon?vel — criar estrutura zerada manualmente

    # Dicion?rio com confrontos definidos
    confrontos_fase_4 = {
        "Jogo 1 (JG1)": ["Gig@ntte", "KING LEONN"],
        "Jogo 2 (JG2)": ["Laranjja Mecannica", "Gr?mio imortal 37"],
        "Jogo 3 (JG3)": ["MauHumor F.C.", "TEAM LOPES 99"],
        "Jogo 4 (JG4)": ["Tabajara de Inhaua FC2", "lsauer fc"]
    }

    # Estrutura zerada para o JS
    classificacao_quartas_sula = {}

    for jogo, times in confrontos_fase_4.items():
        classificacao_quartas_sula[jogo] = []
        for nome_time in times:
            classificacao_quartas_sula[jogo].append({
                "posicao": 0,
                "nome": nome_time,
                "pontos": 0,
                "vitorias": 0,
                "empates": 0,
                "derrotas": 0,
                "totalCartola": 0.0,
                "cartolaSofrido": 0.0,
                "saldoCartola": 0.0
            })

    # Salvar como JS
    with open("2_classificacao_por_jogo_quartas_sula.js", "w", encoding="utf-8") as f:
        f.write("const classificacao_quartas_sula = ")
        json.dump(classificacao_quartas_sula, f, indent=2, ensure_ascii=False)
        f.write(";")

    print("✅ Arquivo JS com classifica??o zerada gerado com sucesso.")


✅ Arquivo JS salvo com sucesso.


In [19]:
def exibir_resultados_rodada(df_rodadas, df_pontuacoes, rodada, jogo=None):
    """
    Exibe os resultados de uma rodada espec?fica, com pontua??o e dados dos times.
    """

    if rodada not in df_rodadas["Rodada"].values:
        return pd.DataFrame([{
            "Jogo": jogo or "-",
            "Rodada": rodada,
            "Mandante_Nome": "-",          
            "Mandante_Pontos": "-",
            "Visitante_Nome": "-",           
            "Visitante_Pontos": "-",
        }])

    df_filtrado = df_rodadas[df_rodadas["Rodada"] == rodada]
    if jogo:
        df_filtrado = df_filtrado[df_filtrado["Jogo"] == jogo]

    resultados = []

    for _, row in df_filtrado.iterrows():
        jogo_ = row["Jogo"]
        mandante = row["Mandante_Nome"]
        visitante = row["Visitante_Nome"]

        pontos_mandante = df_pontuacoes.get(f"Rodada {rodada}", {}).get(mandante, None)
        pontos_visitante = df_pontuacoes.get(f"Rodada {rodada}", {}).get(visitante, None)

        resultados.append({
            "Jogo": jogo_,
            "Rodada": rodada,
            "Mandante_Nome": mandante,
            "Mandante_Pontos": pontos_mandante,
            "Visitante_Nome": visitante,
            "Visitante_Pontos": pontos_visitante
        })

    return pd.DataFrame(resultados)


In [20]:
# Exibir resultados da 15ª rodada
df_resultados_rodada_15 = exibir_resultados_rodada(df_rodadas, df_pontuacoes, rodada=15)

# Exibir apenas os resultados do Grupo B na 1? rodada
df_resultados_jogo_1 = exibir_resultados_rodada(df_rodadas, df_pontuacoes, rodada=15, jogo="Jogo 1 (JG1)")

# Exibir
display(df_resultados_rodada_15)

,Jogo,Rodada,Mandante_Nome,Mandante_Pontos,Visitante_Nome,Visitante_Pontos
0,Jogo 1 (JG1),15,JUV. KP,55.05,Máquina Laranjja,37.83
1,Jogo 2 (JG2),15,dasdoresfc,55.90,JV5 Tricolor Gaúcho,55.90
2,Jogo 3 (JG3),15,FÚRIA LEON,43.73,AZURRA82,43.82
3,Jogo 4 (JG4),15,Grêmio imortal 37,67.55,Mau Humor F.C.,68.35


In [21]:
# Criar arquivo com uma aba para cada rodada contendo os resultados detalhados
from pathlib import Path

# Caminho do arquivo de sa?da
caminho_resultados = "2_resultados_quartas_sula.xlsx"

# Descobrir as rodadas ?nicas no DataFrame
rodadas_disponiveis = sorted(df_rodadas["Rodada"].unique())

with pd.ExcelWriter(caminho_resultados) as writer:
    for rodada in rodadas_disponiveis:
        df_resultados = exibir_resultados_rodada(df_rodadas, df_pontuacoes, rodada=rodada)
        nome_aba = f"Rodada {rodada}"
        df_resultados.to_excel(writer, sheet_name=nome_aba, index=False)

print(f"Arquivo salvo com sucesso: {Path(caminho_resultados).resolve()}")

if "df_resultados" in locals():
    display(df_resultados)
else:
    display(pd.DataFrame())

Arquivo salvo com sucesso: C:\Users\ferna\Projetos\GitHub\cartola_2026\sulamericana\datasets_sula\2_resultados_quartas_sula.xlsx


,Jogo,Rodada,Mandante_Nome,Mandante_Pontos,Visitante_Nome,Visitante_Pontos
0,Jogo 1 (JG1),16,Máquina Laranjja,56.93,JUV. KP,75.88
1,Jogo 2 (JG2),16,JV5 Tricolor Gaúcho,103.39,dasdoresfc,82.25
2,Jogo 3 (JG3),16,AZURRA82,59.61,FÚRIA LEON,66.58
3,Jogo 4 (JG4),16,Mau Humor F.C.,74.28,Grêmio imortal 37,88.68


In [22]:
resultados_js = []

for rodada in sorted(df_rodadas["Rodada"].unique()):
    df_resultados = exibir_resultados_rodada(df_rodadas, df_pontuacoes, rodada=rodada)
    
    for _, row in df_resultados.iterrows():
        resultado = {
            "jogo": row["Jogo"],
            "rodada": int(rodada),
            "mandante": {
                "nome": row["Mandante_Nome"],
                "pontos": float(row["Mandante_Pontos"]) if row["Mandante_Pontos"] is not None else None
            },
            "visitante": {
                "nome": row["Visitante_Nome"],
                "pontos": float(row["Visitante_Pontos"]) if row["Visitante_Pontos"] is not None else None
            },
            "vencedor": (
                "mandante" if row["Mandante_Pontos"] is not None and row["Visitante_Pontos"] is not None and row["Mandante_Pontos"] > row["Visitante_Pontos"]
                else "visitante" if row["Mandante_Pontos"] is not None and row["Visitante_Pontos"] is not None and row["Mandante_Pontos"] < row["Visitante_Pontos"]
                else "empate" if row["Mandante_Pontos"] == row["Visitante_Pontos"] and row["Mandante_Pontos"] is not None
                else "indefinido"
            )

        }
        resultados_js.append(resultado)

# Exportar para arquivo .js
import json

with open("2_resultados_quartas_sula.js", "w", encoding="utf-8") as f:
    f.write("const resultados_quartas_sula = ")
    f.write(json.dumps(resultados_js, indent=2, ensure_ascii=False))
    f.write(";\n")

try:
    rodada_ref = int(rod_ref)
except Exception:
    rodada_ref = 0

parcial_payload = {"rodada": rodada_ref, "times": {}}
try:
    col_parcial = f"Rodada {rodada_ref}"
    if col_parcial in df_pontuacoes.columns:
        times_map = {}
        for nome in df_pontuacoes.index:
            if nome not in ids_times:
                continue
            try:
                val = df_pontuacoes.at[nome, col_parcial]
            except Exception:
                continue
            if str(val) in ("", "nan"):
                continue
            try:
                times_map[str(ids_times[nome])] = float(val)
            except Exception:
                continue
        parcial_payload["times"] = times_map
except Exception:
    pass

try:
    status_http_val = int(status_http)
except Exception:
    status_http_val = None

sula_meta = {
    "rodada_atual": rodada_ref,
    "parcial_disponivel": bool(status_http_val == 2 and parcial_payload["times"])
}

with open("2_resultados_quartas_sula.js", "a", encoding="utf-8") as f:
    f.write("const pontuacaoParcialRodadaAtual = ")
    f.write(json.dumps(parcial_payload, indent=2, ensure_ascii=False))
    f.write(";\n")
    f.write("window.sulaMeta = ")
    f.write(json.dumps(sula_meta, indent=2, ensure_ascii=False))
    f.write(";\n")


### Identificando Vencedores das Quartas de Final da Sul Americana

In [23]:
def obter_classificados_com_id(resultados, ids_por_nome):
    classificados = []
    jogos_agrupados = {}
    for jogo in resultados:
        chave = jogo['jogo']
        if chave not in jogos_agrupados:
            jogos_agrupados[chave] = []
        jogos_agrupados[chave].append(jogo)

    for chave, partidas in jogos_agrupados.items():
        if len(partidas) < 2:
            continue

        partidas = sorted(partidas, key=lambda x: x['rodada'])
        ida, volta = partidas

        if ida['mandante']['pontos'] is None or volta['mandante']['pontos'] is None:
            continue

        time1 = ida['mandante']['nome']
        time2 = ida['visitante']['nome']

        pontos_time1 = ida['mandante']['pontos'] + volta['visitante']['pontos']
        pontos_time2 = ida['visitante']['pontos'] + volta['mandante']['pontos']

        if pontos_time1 > pontos_time2:
            vencedor = time1
        elif pontos_time2 > pontos_time1:
            vencedor = time2
        else:
            vencedor = "EMPATE"

        classificados.append({
            "jogo": chave,
            "classificado_nome": vencedor,
            "classificado_id": ids_por_nome.get(vencedor) if vencedor in ids_por_nome else None
        })

    return classificados


In [24]:
# Inverter os nomes
ids_por_nome = {v: k for k, v in nomes_por_id.items()}

# Ler o resultados_quartas_sula.js (como j? fizemos antes)
with open("2_resultados_quartas_sula.js", "r", encoding="utf-8") as f:
    conteudo = f.read()

conteudo_json = conteudo.replace("const resultados_quartas_sula = ", "", 1)
conteudo_json = conteudo_json.split("\nconst pontuacaoParcialRodadaAtual = ", 1)[0].strip().rstrip(";")
resultados = json.loads(conteudo_json)

# Obter os classificados
classificados = obter_classificados_com_id(resultados, ids_por_nome)

# Salvar
df_classificados = pd.DataFrame(classificados)
df_classificados.to_excel("2_classificados_quartas_sula.xlsx", index=False)

with open("2_classificados_quartas_sula.js", "w", encoding="utf-8") as f:
    f.write("const classificados_quartas_sula = ")
    json.dump(classificados, f, ensure_ascii=False, indent=2)
    f.write(";")
print("✅ Classificados salvos com sucesso em '2_classificados_quartas_sula.xlsx' e '2_classificados_quartas_sula.js'.")

# Exibir para validar
print("Classificados encontrados:", classificados)


✅ Classificados salvos com sucesso em '2_classificados_quartas_sula.xlsx' e '2_classificados_quartas_sula.js'.
Classificados encontrados: [{'jogo': 'Jogo 1 (JG1)', 'classificado_nome': 'JUV. KP', 'classificado_id': 13951133}, {'jogo': 'Jogo 2 (JG2)', 'classificado_nome': 'JV5 Tricolor Gaúcho', 'classificado_id': 1747619}, {'jogo': 'Jogo 3 (JG3)', 'classificado_nome': 'FÚRIA LEON', 'classificado_id': 18344271}, {'jogo': 'Jogo 4 (JG4)', 'classificado_nome': 'Grêmio imortal 37', 'classificado_id': 24468241}]
